surefire-providers/surefire-junit-platform/src/main/java/org/apache/maven/surefire/junitplatform/JUnitPlatformProvider.java

```diff
@@ -172,25 +171,13 @@ private void invokeAllTests(TestsToRun testsToRun, RunListenerAdapter adapter) {
     }
 
     private void execute(TestsToRun testsToRun, RunListenerAdapter adapter) {
-        if (testsToRun.allowEagerReading()) {
-            List<DiscoverySelector> selectors = new ArrayList<>();
-            testsToRun.iterator().forEachRemaining(c -> selectors.add(selectClass(c.getName())));
-
+        testsToRun.iterator().forEachRemaining(c -> {
             LauncherDiscoveryRequestBuilder builder = request()
                     .filters(filters)
                     .configurationParameters(configurationParameters)
-                    .selectors(selectors);
-
+                    .selectors(selectClass(c.getName()));
             launcher.execute(builder.build(), adapter);
-        } else {
```

Comment:
How this has been collapsed and both cases are identical. Can you explain why this new code is correct for both? I honest cannot assess whether this is wrong or correct. What should happen in your opinion in the eager case?

Reply:
Thinking about this a bit, I think I am misunderstanding what the eager reading actually means. Could you please clarify its intention? How does an end user set it and where?

In my case here, what is happening is, in a parallel run, the test runs are combined due to eager reading, the classes get executed together and hence tests get reported as a bundle (I'll look into the reporting aspect, maybe that's the issue and not the execution). But in general, I felt there is no need for eager reading for a parallel run. Maybe we need to separate the 2 conditions?

Reply:
> Thinking about this a bit, I think I am misunderstanding what the eager reading actually means. Could you please clarify its intention? How does an end user set it and where?

Unfortunately, I cannot answer these question because I have no idea.

> In my case here, what is happening is, in a parallel run, the test runs are combined due to eager reading, the classes get executed together and hence tests get reported as a bundle (I'll look into the reporting aspect, maybe that's the issue and not the execution). But in general, I felt there is no need for eager reading for a parallel run. Maybe we need to separate the 2 conditions?

Is this similar: https://issues.apache.org/jira/browse/SUREFIRE-2195?

Reply:
Yes, 2195 seems like the same issue

Reply:
With this change I see in logs:

```
[INFO] Running pkg.domain.BxTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.014 s -- in pkg.domain.BxTest
[INFO] pkg.domain.BxTest.test -- Time elapsed: 0.007 s
[INFO] Running pkg.domain.AxTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s -- in pkg.domain.AxTest
[INFO] pkg.domain.AxTest.test -- Time elapsed: 0 s
```

but without:

```
[INFO] Running pkg.domain.AxTest
[INFO] Running pkg.domain.BxTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.015 s -- in pkg.domain.AxTest
[INFO] pkg.domain.BxTest.test -- Time elapsed: 0.007 s
[INFO] pkg.domain.AxTest.test -- Time elapsed: 0.007 s
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.015 s -- in pkg.domain.BxTest
```

so look like tests are executed sequentially not in parallel

So for me problem is something else ...

Reply:
The issue seems to be in the test listener which isn't parallel safe and is expecting results to trickle in sequentially, unsure of what needs to be fixed.

Reply:
I linked some of similar issues ... the problem is probbably in StatelessXmlReporter

Reply:
@slawekjaranowski Do you think you can pick up, I have honestly no clue here.

Reply:
One alternate approach I am investigating is to ignore these xmls and use [Allure](https://allurereport.org/docs/junit5/) reports instead. They generate a bunch of json files which seem accurate and have all the relevant information.